src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+                    )
+                })?
+                .try_into()
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!(
+                            "The value for pack index checksum on line {} is not the right length",
+                            line_no
+                        ),
+                        None,
+                    )
+                })?;
+
+            let pack_checksum = hex::decode(pack_checksum)
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!("Bad pack checksum format on line {}", line_no),
+                        None,
+                    )
+                })?
+                .try_into()
+                .map_err(|_| {
+                    Error::BadRemoteIndexFormat(
+                        format!(
+                            "The value for pack checksum on line {} is not the right length",
+                            line_no
+                        ),
+                        None,
+                    )
+                })?;
+
+            packs.push(RemotePack {
+                url: absolute_url.as_str().to_owned(),
+                index_checksum,
+                pack_checksum,
+            });
+        }
+
+        Ok(Self {
+            path: None,
+            meta,
+            url,
+            packs,
+        })
+    }
+
+    fn read_keyed_line<R: BufRead>(lines: &mut io::Lines<R>, key: &str) -> Result<String, Error> {
+        let line = match lines.next() {
+            None => Err(Error::BadRemoteIndexFormat(
+                format!("Expected '{} ...', but end of file was reached!", key),
+                None,
+            )),
+            Some(Err(e)) => Err(Error::IOError(e)),
+            Some(Ok(line)) => Ok(line),
+        }?;
+
+        let mut parts = line
+            .split(Self::is_field_separator)
+            .filter(|&p| !p.is_empty());
+
+        if parts.next() != Some(key) {
+            return Err(Error::BadRemoteIndexFormat(
+                format!("Expected '{} ...': '{}'", key, line),
+                None,
+            ));
+        }
+
+        let value = parts.next().ok_or_else(|| {
+            Error::BadRemoteIndexFormat(
+                format!("Expected a value after '{}': '{}'", key, line),
+                None,
+            )
+        })?;
+
+        if parts.next().is_some() {
+            return Err(Error::BadRemoteIndexFormat(
+                format!("Too many values: '{}'", line),
+                None,
+            ));
+        }
+
+        Ok(value.to_string())
+    }
+
+    fn is_field_separator(ch: char) -> bool {
+        ch == ' ' || ch == '\t'
+    }
+}
+
+impl std::fmt::Display for RemoteIndex {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if let Some(stem) = self.path().and_then(|p| p.file_stem()) {
+            write!(fmt, "{} ({})", stem.to_string_lossy(), self.url)?;
```

Comment:
Is it necessary to use `.to_string_lossy()` here as opposed to `.display()`?

Reply:
No, I guess it's not.

Reply:
Actually, it is, apparently `OsStr` does not implement `Display`.